File: infobip-mobile-messaging-android-chat-sdk/src/main/java/org/infobip/mobile/messaging/chat/view/InAppChatWebView.java
package org.infobip.mobile.messaging.chat.view;


import android.annotation.SuppressLint;
import android.content.Context;
import android.net.Uri;
import android.util.AttributeSet;
import android.webkit.WebSettings;
import android.webkit.WebView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.infobip.mobile.messaging.chat.core.InAppChatMobileImpl;
import org.infobip.mobile.messaging.chat.core.InAppChatWebChromeClient;
import org.infobip.mobile.messaging.chat.core.InAppChatWebViewClient;
import org.infobip.mobile.messaging.chat.core.InAppChatWebViewManager;
import org.infobip.mobile.messaging.util.ResourceLoader;
import org.infobip.mobile.messaging.util.StringUtils;

public class InAppChatWebView extends WebView {
    private static final String IN_APP_CHAT_MOBILE_INTERFACE = "InAppChatMobile";
    private static final String RES_ID_IN_APP_CHAT_WIDGET_URI = "ib_inappchat_widget_uri";

    private String widgetUri;

    public InAppChatWebView(Context context) {
        super(context);
    }

    public InAppChatWebView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public InAppChatWebView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @SuppressLint({"AddJavascriptInterface", "SetJavaScriptEnabled"})
    public void setup(InAppChatWebViewManager webViewManager) {
        widgetUri = ResourceLoader.loadStringResourceByName(getContext(), RES_ID_IN_APP_CHAT_WIDGET_URI);
        WebSettings webViewSettings = getSettings();
        webViewSettings.setJavaScriptEnabled(true);
        webViewSettings.setDomStorageEnabled(true);
        webViewSettings.setDatabaseEnabled(true);
        setClickable(true);
        setWebViewClient(new InAppChatWebViewClient(webViewManager));
        addJavascriptInterface(new InAppChatMobileImpl(webViewManager), IN_APP_CHAT_MOBILE_INTERFACE);
        setWebChromeClient(new InAppChatWebChromeClient());
    }

    public void loadChatPage(
            @NonNull String pushRegistrationId,
            @NonNull String widgetId,
            @Nullable String jwt,
            @Nullable String domain
    ) {
        Uri.Builder builder = new Uri.Builder()
                .encodedPath(widgetUri)
                .appendQueryParameter("pushRegId", pushRegistrationId)
                .appendQueryParameter("widgetId", widgetId);

        if (StringUtils.isNotBlank(jwt)) {
            builder.appendQueryParameter("jwt", jwt);
        }

        if (StringUtils.isNotBlank(domain)) {
            builder.appendQueryParameter("domain", domain);
        }

        String resultUrl = builder.build().toString();
        loadUrl(resultUrl);
    }

}

File: infobip-mobile-messaging-android-chat-sdk/src/main/java/org/infobip/mobile/messaging/chat/view/InAppChatWebView.kt
package org.infobip.mobile.messaging.chat.view

import android.annotation.SuppressLint
import android.content.Context
import android.net.Uri
import android.util.AttributeSet
import android.webkit.WebSettings.LOAD_NO_CACHE
import android.webkit.WebView
import org.infobip.mobile.messaging.chat.R
import org.infobip.mobile.messaging.chat.core.InAppChatMobileImpl
import org.infobip.mobile.messaging.chat.core.InAppChatWebChromeClient
import org.infobip.mobile.messaging.chat.core.InAppChatWebViewClient
import org.infobip.mobile.messaging.chat.core.InAppChatWebViewManager
import org.infobip.mobile.messaging.logging.MobileMessagingLogger

@SuppressLint("SetJavaScriptEnabled")
internal class InAppChatWebView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
    defStyleRes: Int = 0
) : WebView(context, attrs, defStyleAttr, defStyleRes) {

    companion object {
        private const val TAG = "InAppChatWebView"
        private const val IN_APP_CHAT_MOBILE_INTERFACE = "InAppChatMobile"
        private const val DEFAULT_WIDGET_URI = "https://livechat.infobip.com/widget.js"
    }

    private val widgetPageUri = context.getString(R.string.ib_inappchat_widget_page_uri)
    private val widgetUri = context.getString(R.string.ib_inappchat_widget_uri)

    init {
        settings.apply {
            javaScriptEnabled = true
            domStorageEnabled = true
            cacheMode = LOAD_NO_CACHE
        }
        isClickable = true
        webChromeClient = InAppChatWebChromeClient()
    }

    fun setup(webViewManager: InAppChatWebViewManager) {
        webViewClient = InAppChatWebViewClient(webViewManager)
        addJavascriptInterface(InAppChatMobileImpl(webViewManager), IN_APP_CHAT_MOBILE_INTERFACE)
    }

    fun loadChatPage(
        pushRegistrationId: String,
        widgetId: String,
        jwt: String? = null,
        domain: String? = null,
        widgetTheme: String? = null,
    ) {
        val builder = Uri.Builder()
            .encodedPath(widgetPageUri)
            .appendQueryParameter("pushRegId", pushRegistrationId)
            .appendQueryParameter("widgetId", widgetId)

        val encodedWidgetUri = widgetUri.takeIf { it != DEFAULT_WIDGET_URI }?.encode()

        if (encodedWidgetUri?.isNotBlank() == true) {
            builder.appendQueryParameter("widgetUri", encodedWidgetUri)
        }

        if (jwt?.isNotBlank() == true) {
            builder.appendQueryParameter("jwt", jwt)
        }

        if (domain?.isNotBlank() == true) {
            builder.appendQueryParameter("domain", domain)
        }

        if (widgetTheme?.isNotBlank() == true) {
            builder.appendQueryParameter("widgetTheme", widgetTheme)
        }

        val resultUrl = builder.build().toString()
        MobileMessagingLogger.d(TAG, "Loading page: $resultUrl")
        loadUrl(resultUrl)
    }

    private fun String.encode(): String = Uri.encode(this)

}